Cabal/Distribution/Compat/Lens.hs

```diff
@@ -0,0 +1,140 @@
+{-# LANGUAGE RankNTypes #-}
+-- | This module provides very basic lens functionality, without extra dependencies.
+--
+-- For the documentation of the combinators see <http://hackage.haskell.org/package/lens lens> package.
+-- This module uses the same vocabulary.
+module Distribution.Compat.Lens (
+    -- * Types
+    Lens,
+    Lens',
+    Traversal,
+    Traversal',
+    -- ** rank-1 types
+    Getting,
+    ASetter,
+    -- * Folds
+    toDListOf,
+    toListOf,
+    toSetOf,
+    -- * Common lenses
+    _2,
+    -- * Cabal developer info
+    -- $development
+    ) where
+
+import Prelude()
+import Distribution.Compat.Prelude
+
+import Control.Applicative (Const (..))
+import Data.Functor.Identity (Identity (..))
+
+import qualified Distribution.Compat.DList as DList
+import qualified Data.Set as Set
+
+-------------------------------------------------------------------------------
+-- Types
+-------------------------------------------------------------------------------
+
+type Lens      s t a b = forall f. Functor f     => (a -> f b) -> s -> f t
+type Traversal s t a b = forall f. Applicative f => (a -> f b) -> s -> f t
+
+type Lens'      s a = forall f. Functor f     => (a -> f a) -> s -> f s
+type Traversal' s a = forall f. Applicative f => (a -> f a) -> s -> f s
+
+type Getting r s a = (a -> Const r a) -> s -> Const r s
+type ASetter s t a b = (a -> Identity b) -> s -> Identity t
+
+-------------------------------------------------------------------------------
+-- Folds
+-------------------------------------------------------------------------------
+
+toDListOf :: Getting (DList.DList a) s a -> s -> DList.DList a
+toDListOf l s = getConst (l (\x -> Const (DList.singleton x)) s)
+
+toListOf :: Getting (DList.DList a) s a -> s -> [a]
+toListOf l = DList.runDList . toDListOf l
+
+toSetOf  :: Getting (Set.Set a) s a -> s -> Set.Set a
+toSetOf l s = getConst (l (\x -> Const (Set.singleton x)) s)
+
+-------------------------------------------------------------------------------
+-- Common
+-------------------------------------------------------------------------------
+
+_2 ::  Functor f => (a -> f b) -> (c, a) -> f (c, b)
```

Comment:
This is btw a good question. Should we introduce `FieldN` classes for these, like `lens` and `microlens` do? (I have `_4` used somewhere already)